### ISSUE TITLE
Update supported Python versions

### ISSUE THREAD

**Issue by asoplata** [jonescompneurolab/hnn-core#947]
Currently, we allow installation in all Python versions 3.8 and up, but there are several changes that need to be made:

- [ ] NEURON currently does not support the newly-released stable version Python 3.13 (see https://github.com/neuronsimulator/nrn/issues/2640 ). They probably will soon, but until that happens, we should not allow installation with 3.13.
- [ ] We only test against 3.8, 3.11, and 3.12, but not 3.9 nor 3.10. I have done local tests against 3.9 and 3.10 on Windows, Mac, and Linux, and I can't find any reason to include those versions in our tests.
- [ ] We should add "classifiers" for all the versions we support into our `setup.py` (example: https://github.com/NeuralEnsemble/neurotune/pull/32/commits/e8dadb5ac43083883cc1bc72c76e7205a275fe59 )

**Comment by jasmainak** [comment]
> I have done local tests against 3.9 and 3.10 on Windows, Mac, and Linux, and I can't find any reason to include those versions in our tests.

save the environment :) running these tests takes computation, if the gain is incremental, it's probably not worth it.

**Comment by asoplata** [comment]
me considering this and all its ramifications
![image](https://github.com/user-attachments/assets/cf44626d-d28c-4448-afe4-695e2fe15929)

You raise an excellent point. To me, this begs the question: in addition to not expanding our CI runs, why not also go further and decrease the CI runs we run by default? Now that we have a full-time developer, I think an effective setup could be the following:

- For all new PRs and their new commits, we run:
    - The current CI runners for MacOS, Linux, and Windows, but only for one major version, such as 3.12, which I believe is the current default version used by Anaconda (which we recommend using).

- Separately, we make a new Github Action to run our CI runners for `master` across all supported Python versions, but only once or twice a month. The vast, vast majority of our bugs are not version incompatibilities, but if any do arise, then I can use these tests (or my own local cross-version tests) to catch them.

- Another idea, depending on difficulty, would be to split the Sphinx doc generation runner into 2 parts:
    1. First part is the actual API documentation, autogenerated from docstrings etc., and whose runner could continue to run for every PR
    2. Second part would be the CPU-expensive Examples and Tutorials documentation generation. Similarly to above, this could be run infrequently, such as part of the once or twice a month CI. I don't think these necessarily need to be re-done for every PR, although I have yet to dive into understanding CircleCI's caching. If we don't already, we should have some tests which mock what is required for all the function calls in the Examples and Tutorials of the documentation.

**Comment by jasmainak** [comment]
> would be to split the Sphinx doc generation runner into 2 parts:

see `make html` and `make html-noplot`. They essentially do this. I think mne-python has some CI magic to detect which files have been touched and only re-build those. Essentially after each CI build, you would save the build to a cache. Then, before the build the cache would be restored. If sphinx-gallery is asked to re-build in this directory, it will only re-build files that have actually changed.

**Comment by asoplata** [comment]
Adding a note about this: we may want to drop "official" support for 3.8, including for testing, since it is officially no longer supported (see https://devguide.python.org/versions/ ). From personal recollection, I feel like I have noticed `macos-3.8` failing the test a bit more frequently than other versions, though I have no data to back this up. 

That said, if someone reaches out to us and specifically requests 3.8 support, we can totally do that, either entirely or (if there are growing incompatibilities) in a special branch.

**Comment by jasmainak** [comment]
It's too complicated to support a special version in a separate branch :) I would recommend just dropping support when we are ready ... I am one of those guilty users still using 3.8, should switch to latest version!

**Comment by jasmainak** [comment]
Virtual environments exist for someone wanting to use older version. But it would be hard to provide continued support due to the very reasons you mention + small dev team

**Comment by asoplata** [comment]
For better or worse, it appears that, until NEURON works for Python 3.13, our current install will always fail for 3.13. However, the failure message is not very helpful to a general user, and does not communicate the "real" install problem -- below is what I get if I try to install in a 3.13 environment:

```
(test) {13:48}~/Desktop/hnn-core:master ✗ ➭ python -m pip install .
Processing /Users/austinsoplata/Desktop/hnn-core
  Installing build dependencies ... error
  error: subprocess-exited-with-error

  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [4 lines of output]
      Collecting setuptools>=40.8.0
        Using cached setuptools-76.0.0-py3-none-any.whl.metadata (6.7 kB)
      ERROR: Could not find a version that satisfies the requirement NEURON>=7.7 (from versions: none)
      ERROR: No matching distribution found for NEURON>=7.7
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

This is the output that always occurs when I try to install in a 3.13 environment, *regardless of any Python-version changes or constraints I add to `pyproject.toml` or `setup.py`*. The "real" install problem here is that when it says `Could not find a version...` it means that it couldn't find a version of Python that works for `NEURON>=7.7`, since NEURON currently doesn't allow installation on Python 3.13. The reason that this error shows despite any Python version restrictions elsewhere is I believe appears that the `pyproject.toml` `build-system` requirements (see https://github.com/jonescompneurolab/hnn-core/blob/master/pyproject.toml#L2 ) are installed before any other part of the installation, including later Python version requirements.

In other words, even if, in our `setup.py`, we change `python_requires='>=3.8'` to `python_requires='>=3.8, < 3.13'`, that `python_requires` check is not run BEFORE the above `pyproject.toml` `build-system` requirements. Therefore, the above error shown is still what gets output.

Of course, I tried adding `python<3.13` etc. to the `pyproject.toml` `build-system` requirements, but checking the python version that way does not appear to be supported; I was getting failures in every case.

There *may* be a way to print a more user-friendly error if we use `build_meta` as shown in https://setuptools.pypa.io/en/latest/build_meta.html#dynamic-build-dependencies-and-other-build-meta-tweaks but I think this is a terrible idea for many reasons. One reason in particular is any "fixes" we do will be obviated by whenever NEURON releases a 3.13 compatible version, which is probably not far off.

Given the above, **I propose the following**: We still change `setup.py`'s `python_requires='>=3.8'` to `python_requires='>=3.8, < 3.13'` and add the Classifiers like above, and we explicitly add mention of not currently supporting 3.13 to all install documentation. It is possible (I'm not sure) that if we next upload a version with these python version constraints to `pip`, that `pip` itself may provide a better version-mismatch error to the user before they try to install `NEURON>=7.7`; it certainly won't hurt.

**Comment by asoplata** [comment]
Moving forward: the next step is solely to add documentation to our README and Installation that one should not use Python 3.13 for now.

**Comment by asoplata** [comment]
This will be moved to Milestone v0.5, because the changes to the Github Workflows will be non-trivial, but not needed for the April Workshop.
